processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java

```diff
@@ -184,7 +187,18 @@ public boolean matches()
           return matchNull;
         } else {
           for (int i = 0; i < size; ++i) {
-            if (predicateMatchingValueIds.get(row.get(i))) {
+            final int id = row.get(i);
+            final boolean matches;
+
+            if (checkedIds.get(id)) {
+              matches = matchingIds.get(id);
+            } else {
+              matches = predicate.apply(selector.lookupName(id));
+              checkedIds.set(id);
+              matchingIds.set(id, matches);
```

Comment:
`if (matches) { matchingIds.set(id); }` is better

Reply:
This transformation gives worse performance in the benchmark:

```
FilterPartitionBenchmark.readWithExFnPostFilter            750000     basic  avgt   25  10587.515 ± 789.090  us/op
```

I suspect it is due to adding a branch where none previously existed.

Reply:
It's impossible because if you look at the source code of set(index, boolean), it has the same branching, plus it does extra work in the case when `matches` is false, that we can avoid. Something is wrong with your measurements.

Reply:
Well, you're right. Even though I tried it twice yesterday (since I thought it surprising) and got this result, this morning I tried again and got something that was in line with the original result (around 8800 µs/op). So I'll make this minor transformation, because why not. The other, bigger transformation (collapsing to one bitset) still measures as slower though.

Reply:
Pushed the minor transformation.